jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -559,4 +574,222 @@ public String toString()
+    {
+        private final ThreadLocal<List<Pool<T>.Entry>> cache;
+        private final int cacheSize;
+
+        ThreadLocalListCache(int size)
+        {
+            this.cacheSize = size;
+            this.cache = ThreadLocal.withInitial(() -> new ArrayList<>(cacheSize));
+        }
+
+        @Override
+        public Pool<T>.Entry acquire(List<Pool<T>.Entry> entries)
+        {
+            List<Pool<T>.Entry> cachedList = cache.get();
+            while (!cachedList.isEmpty())
+            {
+                Pool<T>.Entry cachedEntry = cachedList.remove(cachedList.size() - 1);
+                if (cachedEntry.tryAcquire())
+                    return cachedEntry;
+            }
+            return null;
+        }
+
+        @Override
+        public void released(List<Pool<T>.Entry> entries, Pool<T>.Entry entry, boolean reusable)
+        {
+            if (reusable)
+            {
+                List<Pool<T>.Entry> cachedList = cache.get();
+                if (cachedList.size() < cacheSize)
+                    cachedList.add(entry);
+            }
+        }
+    }
+
+    private abstract static class IndexedCached<T> implements Cache<T>
+    {
+        @Override
+        public Pool<T>.Entry acquire(List<Pool<T>.Entry> entries)
+        {
+            int size = entries.size();
+            if (size == 0)
+                return null;
+            int i = nextIndex(size);
+            try
+            {
+                Pool<T>.Entry entry = entries.get(i);
+                if (entry != null && entry.tryAcquire())
+                    return entry;
+            }
+            catch (Exception e)
+            {
+                // Could be out of bounds
+                LOGGER.ignore(e);
+            }
+            return null;
+        }
+
+        protected abstract int nextIndex(int size);
+    }
+
+    public static class RandomCache<T> extends IndexedCached<T>
+    {
+        @Override
+        protected int nextIndex(int size)
+        {
+            return ThreadLocalRandom.current().nextInt(size);
+        }
+    }
+
+    public static class RoundRobinCache<T> extends IndexedCached<T>
+    {
+        AtomicInteger index = new AtomicInteger();
+
+        @Override
+        protected int nextIndex(int size)
+        {
+            return index.getAndUpdate(c -> Math.max(0, c + 1)) % size;
+        }
+    }
+
+    public static class RoundRobinStrategy<T> extends RoundRobinCache<T> implements Strategy<T>
+    {
+        @Override
+        public Pool<T>.Entry acquire(List<Pool<T>.Entry> entries)
+        {
+            int tries = entries.size();
+            while (tries-- > 0)
+            {
+                Pool<T>.Entry entry = super.acquire(entries);
+                if (entry != null)
+                    return entry;
+            }
+            return null;
+        }
+    }
+
+    public static class LeastRecentlyUsedStrategy<T> implements Strategy<T>
```

Comment:
do we want to keep this strategy?

Reply:
I'll keep for now to test API during dev... but don't think we should include it in final merge... actually I might move it to a test to check that an external strategy can be applied without access to protected methods